src/sentry/scim/endpoints/utils.py

```diff
@@ -0,0 +1,138 @@
+            return False
+
+        return True
+
+
+class SCIMEndpoint(OrganizationEndpoint):
+    permission_classes = (OrganizationSCIMPermission,)
+    content_negotiation_class = IgnoreClientContentNegotiation
+
+    def paginate(
+        self,
+        request,
+        on_results=None,
+        paginator=None,
+        paginator_cls=GenericOffsetPaginator,
+        default_per_page=100,
+        max_per_page=100,
+        cursor_cls=Cursor,
+        **paginator_kwargs,
+    ):
+        assert (paginator and not paginator_kwargs) or (paginator_cls and paginator_kwargs)
+
+        per_page = self.get_per_page(request, default_per_page, max_per_page)
+
+        input_cursor = None
+        if request.GET.get("startIndex"):
+            # XXX: SCIM startIndex param is 1 indexed
+            try:
+                input_cursor = Cursor(0, int(request.GET.get("startIndex")) - 1, 0)
+            except ValueError:
+                raise ParseError(detail="Invalid cursor parameter.")
+
+        try:
+            with sentry_sdk.start_span(
+                op="base.paginate.get_result",
+                description=type(self).__name__,
+            ) as span:
+                span.set_data("Limit", per_page)
+                cursor_result = paginator.get_result(limit=per_page, cursor=input_cursor)
+        except BadPaginationError as e:
+            raise ParseError(detail=str(e))
+
+        # map results based on callback
+        if on_results:
+            with sentry_sdk.start_span(
+                op="base.paginate.on_results",
+                description=type(self).__name__,
+            ):
+                results = on_results(cursor_result.results)
+        else:
+            results = cursor_result.results
+        context = {
+            "schemas": [SCIM_API_LIST],
+            "totalResults": paginator_kwargs["queryset"].count(),  # TODO: audit perf
+            "startIndex": int(request.GET.get("startIndex", 1)),  # must be integer
+            "itemsPerPage": len(results),  # what's max?
+            "Resources": results,
+        }
+        return Response(context)
+
+
+def parse_filter_conditions(raw_filters):
+    # TODO: support "and" operator
+    # TODO: support email querying/filtering
+    filters = []
+    if raw_filters is None:
+        return filters
+    conditions = raw_filters.split(",")
+
+    for c in conditions:
+        [key, value] = c.split(" eq ")
+        if not key or not value:
+            continue
+
+        key = key.strip()
+        value = value.strip()
+
+        # For USERS: Unique username should always be lowercase
+        if key == "userName":
+            value = value.lower()
+
+        if value[0] == '"' and value[-1] == '"':
+            value = value.replace('"', "")
+        if value[0] == "'" and value[-1] == "'":
+            value = value.replace("'", "")
+        filters.append([key, value])
+    if len(filters) > 0:
+        filter_val = [filters[0][1]]
```

Comment:
yup, I will clean this up and add some comments. SCIM has an [entire grammar](https://datatracker.ietf.org/doc/html/rfc7644#section-3.4.2.2) for filtering , but for our case i'm not sure that we need the whole thing -- okta at least doesn't use any of it except for "eq", and for other IDPs we can probably state we don't support them, so this thing just ends up looking a bit hacky since the only thing we support is an expression like `"userName eq 'josh.ferge'"`. If it comes back that we need more advanced filtering, we'd probably want to use something like
https://github.com/15five/scim2-filter-parser

Reply:
Would it be easier to understand later if we had a basic PEG grammar for this like we do for our own search grammar?

Reply:
to me it seems like it wouldn't be worth it to do a PEG grammar as for a full implementation i think we'd want to use the above package. maybe i'm overestimating the amount of work that would take but from a quick glance at our search grammar i don't want to dive down that rabbit hole right now.